### PR TITLE
[flang2] Fix segmentation faults (#421)

### PR DESCRIPTION
**Pull request by bryanpkc**
Add null checks to solve several obvious crash bugs.

### CODE REVIEW

**Comment by kiranchandramohan** [comment]
@bryanpkc Can we have some tests as well?

**Comment by bryanpkc** [comment]
@kiranchandramohan I have added some tests.

**Comment by kiranchandramohan** [comment]
@bryanpkc Are these crashes reproducible from Fortran source or only when independently running flang1/flang2?

**Comment by bryanpkc** [comment]
@kiranchandramohan These failures are not normally reproducible from Fortran source. The driver usually ensures that flang2 is invoked with correct command line arguments; however a developer might want to run flang2 directly, and would not expect it to crash when invoked without any argument.

We found the second crash by accident when a file system was full and an empty .ilm file got passed to flang2.

**Comment by bryanpkc** [comment]
flang2 should probably fail in the case of an empty input file, but I just found that Fortran preprocessing (`-E`) expects flang2 to succeed in that case. I'm not really sure why flang2 is required when preprocessing; I can look a bit deeper.

**Comment by bryanpkc** [comment]
main.cpp processes two types of input files, an `.ilm` file and an `.stb` file. Both contains multiple sections that are read in loops, which call `upper` to process each section. The loops depends on the EOF condition to signify a successful termination of the loop; the way the functions are structured, `upper` will always return with an "empty section" condition at the end of an unempty file. I have added counters to let the loops distinguish between an empty section and a really empty input file, and make flang2 produce a warning in the latter case, instead of crashing.

File: tools/flang2/flang2exe/main.cpp [hunk 1]
```diff
@@ -1006,6 +1025,10 @@ process_stb_file()
 
   } while (!gbl.eof_flag);
 
+  if (stb_units_seen == 0) {
```
Comment (kiranchandramohan) [inline tools/flang2/flang2exe/main.cpp:1028]: Nit: drop trivial brace.

File: tools/flang2/flang2exe/main.cpp [hunk 2]
```diff
@@ -428,9 +429,18 @@ main(int argc, char *argv[])
...
       break;
 
+    // We cannot reuse gbl.func_count here since gbl.func_count must be
+    // incremented before calling process_input().
+    ilm_units_seen++;
+
   } while (!gbl.eof_flag);
 
-  cg_llvm_end();
+  if (ilm_units_seen == 0) {
```
Comment (kiranchandramohan) [inline tools/flang2/flang2exe/main.cpp:438]: Nit: Drop trivial brace.

File: tools/flang2/flang2exe/main.cpp [hunk 3]
```diff
@@ -428,9 +429,18 @@ main(int argc, char *argv[])
...
+    errwarn((error_code_t)6);
+  }
+
+  if (FTN_HAS_INIT())
```
Comment (kiranchandramohan) [inline tools/flang2/flang2exe/main.cpp:442]: Nit: A comment here on what FTN_HAS_INIT is and when it is set might be useful.

**Review by kiranchandramohan — APPROVED** [review]
LGTM.

> main.cpp processes two types of input files, an `.ilm` file and an `.stb` file. Both contains multiple sections that are read in loops, which call `upper` to process each section.

Now that you understand the flow, can you write this somewhere if you get the time?

**Review by gklimowicz — APPROVED** [review]
Works fine on OpenPower LLVM 9.0 and 10.0.

**Review by shivaramaarao — APPROVED** [review]
Looks good to me.

**Comment by kiranchandramohan** [comment]
@bryanpkc is this ready to merge?

File: tools/flang2/flang2exe/main.cpp (hunk 3 above)
Comment (bryanpkc) [reply tools/flang2/flang2exe/main.cpp:443]: Done.

File: tools/flang2/flang2exe/main.cpp (hunk 1 above)
Comment (bryanpkc) [reply tools/flang2/flang2exe/main.cpp]: Done.

File: tools/flang2/flang2exe/main.cpp (hunk 2 above)
Comment (bryanpkc) [reply tools/flang2/flang2exe/main.cpp]: Done.

**Comment by bryanpkc** [comment]
Yes, @kiranchandramohan, I was just addressing your comments just as your reminder came.

**Pull request merged**